maven-core/src/main/java/org/apache/maven/project/MavenProject.java

```diff
@@ -921,12 +926,20 @@ public void setInjectedProfileIds( String source, List<String> injectedProfileId
      * coordinates.
      *
      * @param artifact the artifact to add or replace.
-     * @throws DuplicateArtifactAttachmentException
+     * @deprecated Please use {@link MavenProjectHelper}
+     * @throws DuplicateArtifactAttachmentException not used anymore but leave it for backward compatibility
      */
     public void addAttachedArtifact( Artifact artifact )
         throws DuplicateArtifactAttachmentException
     {
-        getAttachedArtifacts().add( artifact );
+        // if already there we remove it and add again
+        int index = attachedArtifacts.indexOf( artifact );
+        if ( index > 0 )
+        {
+            LOGGER.warn( "artifact {} already attached, remove previous instance and add again" );
```

Comment:
So we decide to have last wins?

Reply:
Yes because the javadoc says this `Add or replace an artifact` for so many years but never really did it :)

Reply:
Agreed.